ts41/get.d.ts

```diff
@@ -0,0 +1,123 @@
+import {Split} from './utilities';
+import {StringDigit} from '../source/utilities';
+
+/**
+Like @see Get but receives an array of strings as a path parameter.
+*/
+type GetWithPath<ObjectType, Keys extends readonly string[]> =
+	Keys extends []
+	? ObjectType
+	: Keys extends [infer Head, ...infer Tail]
+	? GetWithPath<PropertyOf<ObjectType, Extract<Head, string>>, Extract<Tail, string[]>>
+	: never;
+
+/**
+Splits a dot-prop style path into a tuple comprised of the properties in the path. Handles square-bracket notation.
+
+@example
+```
+ToPath<'foo.bar.baz'>
+//=> ['foo', 'bar', 'baz']
+
+ToPath<'foo[0].bar.baz'>
+//=> ['foo', '0', 'bar', 'baz']
+```
+*/
+type ToPath<S extends string> = Split<FixPathSquareBrackets<S>, '.'>;
+
+/**
+Replaces square-bracketed dot notation with dots, for example, `foo[0].bar` -> `foo.0.bar`.
+*/
+type FixPathSquareBrackets<Path extends string> =
+	Path extends `${infer Head}[${infer Middle}]${infer Tail}`
+	? `${Head}.${Middle}${FixPathSquareBrackets<Tail>}`
+	: Path;
+
+/**
+Returns true if `S` is made up out of `C` repeated 0 or more times.
+
+@example
+```
+ConsistsOnlyOf<'aaa', 'a'> //=> true
+ConsistsOnlyOf<'aBa', 'a'> //=> false
+ConsistsOnlyOf<'', 'a'> //=> true
+```
+*/
+type ConsistsOnlyOf<LongString extends string, Substring extends string> =
+	LongString extends ''
+	? true
+	: LongString extends `${Substring}${infer Tail}`
+	? ConsistsOnlyOf<Tail, Substring>
+  : false;
+
+/**
+Convert a type which may have number keys to one with string keys, making it possible to index using strings retrieved from template types.
+
+@example
+```
+type WithNumbers = {foo: string; 0: boolean};
+type WithStrings = WithStringKeys<WithNumbers>;
+
+type WithNumbersKeys = keyof WithNumbers;
+//=> 'foo' | 0
+type WithStringsKeys = keyof WithStrings;
+//=> 'foo' | '0'
+```
+*/
+type WithStringKeys<ObjectType extends Record<string | number, any>> = {
+	[Key in `${Extract<keyof ObjectType, string | number>}`]: ObjectType[Key]
```

Comment:
How about the following instead? I think it would be simpler to read?

```suggestion
	[Key in keyof ObjectType as `${Key}`]: ObjectType[Key]
```

Reply:
I did something like that at first, but weirdly typescript isn't able to keep track of the fact that `Key` must be a string or number (I'd have thought it would be able to, since `ObjectType extends Record<string | number, any>`:

![image](https://user-images.githubusercontent.com/15040698/101054747-386c9880-3557-11eb-9ddf-283c0c5f39cb.png)